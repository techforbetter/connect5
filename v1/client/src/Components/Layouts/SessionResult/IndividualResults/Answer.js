import React from "react";
import PropTypes from "prop-types";

import { IndividualAnswer, QuestionSpan, MatrixAnswer } from "../StyledComponents";

class Answer extends React.Component {
  render() {
    const { answer } = this.props;
    if (answer.inputType === "checkbox") {
      return (
        <React.Fragment>
          <ul>
            {answer
              && answer.answer.map(item => (
                <li key={Math.random()}>
                  <MatrixAnswer>{item}</MatrixAnswer>
                </li>
              ))}
          </ul>
        </React.Fragment>
      );
    }
    if (answer.inputType === "matrix") {
      const answerArray = Object.entries(answer.answer);
      return (
        <React.Fragment>
          {answer
            && answerArray.map(item => (
              <div key={Math.random()}>
                <MatrixAnswer>
                  <h4>{item[0]}</h4>
                  <p>{item[1]}</p>
                </MatrixAnswer>
              </div>
            ))}
        </React.Fragment>
      );
    }
    return (
      <IndividualAnswer>
        <QuestionSpan>A.</QuestionSpan>
        {answer.answer}
        {answer.inputType === "radiostar" && <span> stars (out of 6)</span>}
      </IndividualAnswer>
    );
  }
}

// const Answer = ({ answer }) => (

// <div>
//   {
//     answer.inputType === "matrix" || answer.inputType === "checkbox"
//       ? (
//         <ul>
//           {answer && answer.answer.map(item => (
//             <li key={Math.random()}>

//               <MatrixAnswer>
//                 {item}
//               </MatrixAnswer>
//             </li>
//           ))}
//         </ul>

//       )
//       : (
//         <IndividualAnswer>
//           <QuestionSpan>
//                 A.
//           </QuestionSpan>
//           {answer.answer}
//           {
//             answer.inputType === "radiostar"
//               && (
//                 <span>
//                   {" "}
//                         stars (out of 6)
//                 </span>
//               )
//           }
//         </IndividualAnswer>
//       )
//   }
// </div>
// );

export default Answer;

Answer.propTypes = {
  answer: PropTypes.shape({
    answer: PropTypes.any.isRequired,
    inputType: PropTypes.string.isRequired,
  }),
};

Answer.defaultProps = {
  answer: {},
};
