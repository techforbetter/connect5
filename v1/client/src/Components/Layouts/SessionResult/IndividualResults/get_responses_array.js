// get array of responses IDs from answers array
const getResponsesArray = (answers) => {
  const responsesArray = [];
  answers.map((answer) => {
    const flag = responsesArray.find(response => answer.response === response);
    return !flag && responsesArray.push(answer.response);
  });
  return responsesArray;
};

export default getResponsesArray;
