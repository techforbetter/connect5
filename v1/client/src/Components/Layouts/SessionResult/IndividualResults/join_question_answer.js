const joinQuestionsWithAnwsers = (questions, answers) => {
  const output = [];

  answers.map((answer) => {
    const mathcedquestion = questions.filter(question => question._id === answer.question);
    const item = answer;
    // eslint-disable-next-line no-unused-expressions
    mathcedquestion[0] && (item.questionText = mathcedquestion[0].questionText)
    && (item.inputType = mathcedquestion[0].inputType);
    return mathcedquestion[0] && output.push(item);
  });
  // slice the first 5 questions from question array
  return output.slice(5 - output.length);
};

export default joinQuestionsWithAnwsers;
