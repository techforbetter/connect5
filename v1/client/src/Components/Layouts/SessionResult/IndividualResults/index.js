import React, { Component, Fragment } from "react";
import axios from "axios";
import ReactRouterPropTypes from "react-router-prop-types";

import Answer from "./Answer";

import {
  IndividualWrapper,
  NavigationWrapper,
  Arrow,
  ResponseCounter,
  QuestionSpan,
  IndividualQuestion,
  Error,
} from "../StyledComponents";

import getResponsesArray from "./get_responses_array";
import joinQuestionsWithAnwsers from "./join_question_answer";

class IndividualResults extends Component {
  static propTypes = {
    history: ReactRouterPropTypes.history,
    match: ReactRouterPropTypes.match,
  };

  static defaultProps = {
    history: null,
    match: null,
  }

  state ={
    joinedAnswers: [],
    answers: [],
  }

  componentDidMount() {
    const { history, match } = this.props;
    // get session responses from backend
    axios.get(`/session/responses/${match.params.sessionId}/${match.params.sessionType}`)

      // save the data comes from backend into the state
      .then(({ data }) => data.map(item => this.setState(item)))
      .then(() => {
        const { answers, questions } = this.state;
        // get an array of responses IDs and store it in the state
        const responses = getResponsesArray(answers);
        this.setState(() => ({ responses, activeIndex: 0 }), () => {
          // join the answer objects with question objects
          const joinedAnswers = joinQuestionsWithAnwsers(questions, answers);

          if (joinedAnswers.length < 1) {
            this.setState({ noResults: true });
          } else {
            this.setState({ joinedAnswers });
          }
        });
      })
      .catch(() => history.push("/server-error"));
  }

  // controll the navigation through the sessions
  handleNextPrev = (e) => {
    const { activeIndex, responses } = this.state;
    const inc = Number(e.target.id);
    let newIndex = 0;
    if (activeIndex + inc >= responses.length) {
      newIndex = 0;
    } else if (activeIndex + inc < 0) {
      newIndex = responses.length - 1;
    } else {
      newIndex = activeIndex + inc;
    }
    this.setState({ activeIndex: newIndex });
  }


  render() {
    const {
      activeIndex, responses, joinedAnswers, noResults,
    } = this.state;
    const responseNumber = activeIndex + 1;
    const activeresonseid = responses && responses[activeIndex];

    return (
      <IndividualWrapper>

        {
          noResults
            ? <Error>This session has no responses yet :(</Error>
            : (
              <div>
                <NavigationWrapper>
                  <Arrow direction="right" id="-1" onClick={this.handleNextPrev} />

                  <ResponseCounter>
                    {responseNumber}
                    {" "}
              OF
                    {" "}
                    {responses && responses.length}
                  </ResponseCounter>

                  <Arrow direction="left" id="+1" onClick={this.handleNextPrev} />
                </NavigationWrapper>

                {joinedAnswers.map(answer => (answer.response === activeresonseid && (
                  <Fragment key={answer._id}>

                    <IndividualQuestion>
                      <QuestionSpan>
                  Q.
                      </QuestionSpan>
                      {answer.questionText}
                      <Answer answer={answer} />
                    </IndividualQuestion>

                  </Fragment>
                )
                ))}
              </div>
            )
        }


      </IndividualWrapper>
    );
  }
}

export default IndividualResults;
