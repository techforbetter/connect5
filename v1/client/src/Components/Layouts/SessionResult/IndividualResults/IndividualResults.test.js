import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter, Route } from "react-router-dom";

import IndividualResults from "./index";

test("snapshot for 'IndividualResults' layout", () => {
  const context = {};

  const component = renderer.create(
    <BrowserRouter context={context}>
      <Route path="/" exact component={IndividualResults} />
    </BrowserRouter>,
  );
  expect(component).toMatchSnapshot();
});
