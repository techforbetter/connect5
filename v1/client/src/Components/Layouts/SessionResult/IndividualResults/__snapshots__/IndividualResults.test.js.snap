// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot for 'IndividualResults' layout 1`] = `
<div
  className="sc-kkGfuU hQUDDl"
>
  <div>
    <div
      className="sc-iAyFgw gmTSaW"
    >
      <div
        className="sc-hSdWYo bJnYyB"
        direction="right"
        id="-1"
        onClick={[Function]}
      />
      <i
        className="sc-eHgmQL gmKbQu"
      >
        NaN
         
        OF
         
      </i>
      <div
        className="sc-hSdWYo jHdbJh"
        direction="left"
        id="+1"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
